python-lib/prologin/matchmaker/matchmaker.py

```diff
@@ -0,0 +1,308 @@
+        # Commit by replacing old match list to avoid race conditions
+        self.next_matches = self.next_matches + matches
+
+    async def cancel_matches(self, old_champions, current_champions):
+        logging.info('Cancelling matches by champions %s', old_champions)
+
+        # Cancel scheduled matches
+        for old_champion in old_champions:
+            for match in self.matches_with_champion[old_champion]:
+                logging.debug('Cancelling %s', match)
+                match.cancelled = True
+            del self.matches_with_champion[old_champion]
+
+        # Cancel already created matches
+        cancelled_matches = self.tournament.matches.exclude(
+            players__in=current_champions
+        )
+        logging.debug('Cancelling %d created matches', len(cancelled_matches))
+        cancelled_matches.update(status='cancelled')
+        for match in cancelled_matches.all():
+            match.remove(self.tournament)
+
+    async def schedule_new_matches(self, known_champions, new_champions):
+        logging.info('Scheduling matches for new champions %s', new_champions)
+
+        new_matches = []
+        # Iterate over new matches
+        for chs in itertools.product(
+            new_champions | known_champions,
+            repeat=django_settings.STECHEC_NPLAYERS,
+        ):
+            if not new_champions & set(chs):
+                # No now champion in match
+                continue
+
+            # Don't fight against yourself
+            ch_ids = [c.id for c in chs]
+            if len(set(ch_ids)) != len(ch_ids):
+                continue
+
+            new_matches.extend(self.get_matches_with(chs))
+
+        random.shuffle(new_matches)
+        self.add_matches(new_matches)
+
+    async def dbwatcher_task(self):
+        """Watches for new champions."""
+        while True:
+            # Query database
+            current_champions = set(self.get_champions_query())
+
+            old_champions = self.champions - current_champions
+            if old_champions:
+                await self.cancel_matches(old_champions, current_champions)
+
+            # Schedule new matches
+            known_champions = self.champions & current_champions
+            new_champions = current_champions - self.champions
+            if new_champions:
+                await self.schedule_new_matches(known_champions, new_champions)
+
+            # Commit set of current champions
+            self.champions = current_champions
+
+            await asyncio.sleep(1)
+
+    async def match_creator_task(self):
+        """Create new matches at predefined rate per second."""
+        while True:
+            if self.next_matches:
+                logging.info('%d matches in queue', len(self.next_matches))
+
+            new_matches = []
+            while (
+                len(new_matches)
+                < self.config["matchmaker"]["matches_per_second"]
```

Comment:
Looks like this is also missing a max size check, currently this can grow indefinitely (at a fixed rate).

Reply:
The max size of `new_matches` is `matches_per_second`. The virtual list that will grow indefinitely is the number of launched matches, which are scheduled just after this `while` loop.

We do need a global limit to how many matches MatchMaker can create, to avoid overloading masternode. I'll add that shortly.